File: Java/Binary-Tree-Level-Order-Traversal-II.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
 
 
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: buttom-up level order a list of lists of integer
     */
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        // write your code here
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if (root == null) return res;
        q.add(root);
        while (!q.isEmpty()) {
            ArrayList<Integer> cur = new ArrayList<Integer>();
            int size = q.size();
            for (int i = 0; i < size; i ++) {
                TreeNode temp = q.poll();
                cur.add(temp.val);
                if (temp.left != null) q.add(temp.left);
                if (temp.right != null) q.add(temp.right);
            }
            res.add(cur);
        }
        Collections.reverse(res);
        return res;
    }
}

File: Cpp/Binary-Tree-Level-Order-Traversal-II.cpp
/**
 * Definition of TreeNode:
 **/
class TreeNode {
public:
    int val;
    TreeNode *left, *right;
    TreeNode(int val) {
        this->val = val;
        this->left = this->right = NULL;
    }
};


class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Level order a list of lists of integer
     */
public:
    vector<vector<int> > levelOrderButtom(TreeNode *root) {
        // write your code here
        vector<vector<int> > vec;
        if (root == NULL) return vec;
        queue<TreeNode*> q1,q2;
        TreeNode *res;
        q1.push(root);
        while (!q1.empty() || !q2.empty()) {
            vector<int> st;
            if (!q1.empty()) {
                while (!q1.empty()) {
                    res = q1.front();
                    if (res->left != NULL) q2.push(res->left);
                    if (res->right != NULL) q2.push(res->right);
                    q1.pop();
                    st.push_back(res->val);
                }
                vec.push_back(st);
            }
            else {
                while (!q2.empty()) {
                    res = q2.front();
                    if (res->left != NULL) q1.push(res->left);
                    if (res->right != NULL) q1.push(res->right);
                    q2.pop();
                    st.push_back(res->val);
                }
                vec.push_back(st);
            }
        }
        reverse(vec.begin(), vec.end());
        return vec;
    }
};
